extension/background.js

```diff
@@ -0,0 +1,230 @@
+
+    console.debug("registering telemetry events");
+    browser.telemetry.registerEvents(TELEMETRY_CATEGORY, {
+      [TELEMETRY_METHOD_ETLD_CHANGE]: {
+        methods: [TELEMETRY_METHOD_ETLD_CHANGE],
+        objects: [TELEMETRY_OBJECT_WEBREQUEST, TELEMETRY_OBJECT_OTHER],
+        extra_keys: ["addonId", "addonVersion", "from", "to"],
+        record_on_release: true,
+      },
+    });
+  }
+
+  async getMatchPatterns() {
+    try {
+      this.matchPatternsMap =
+        await browser.addonsSearchExperiment.getMatchPatterns();
+    } catch (err) {
+      console.error(`failed to retrieve the list of URL patterns: ${err}`);
+      this.matchPatternsMap = {};
+    }
+
+    return this.matchPatternsMap;
+  }
+
+  async monitor() {
+    // If there is already a listener, remove it so that we can re-add one
+    // after. This is because we're using the same listener with different URL
+    // patterns (when the list of search engines changes).
+    if (
+      browser.webRequest.onBeforeRequest.hasListener(this.webRequestHandler)
+    ) {
+      console.debug("removing onBeforeRequest listener");
+      browser.webRequest.onBeforeRequest.removeListener(this.webRequestHandler);
+    }
+
+    if (
+      browser.webRequest.onBeforeRedirect.hasListener(this.webRequestHandler)
+    ) {
+      console.debug("removing onBeforeRedirect listener");
+      browser.webRequest.onBeforeRedirect.removeListener(
+        this.webRequestHandler
+      );
+    }
+
+    // Retrieve the list of URL patterns to monitor with our listener.
+    //
+    // Note: search suggestions are system principal requests, so webRequest
+    // cannot intercept them.
+    const matchPatternsMap = await this.getMatchPatterns();
+    const patterns = Object.keys(matchPatternsMap);
+
+    if (patterns.length === 0) {
+      console.debug(
+        "not registering any listener because there is no URL to monitor"
+      );
+      return;
+    }
+
+    console.debug("registering onBeforeRequest listener");
+    browser.webRequest.onBeforeRequest.addListener(
+      this.webRequestHandler,
+      { urls: patterns },
+      ["blocking"]
+    );
+
+    // This one is needed in addition to `onBeforeRequest` because this
+    // extension might be registered before or after some extensions.
+    // Depending on that, our `onBeforeRequest` listener might not be called,
+    // which is why we also listen to `onBeforeRedirect`.
+    console.debug("registering onBeforeRedirect listener");
+    browser.webRequest.onBeforeRedirect.addListener(this.webRequestHandler, {
+      urls: patterns,
+    });
+  }
+
+  // `redirectUrl` is usually valid when the redirect has been detected via
+  // `onBeforeRedirect`, otherwise it's likely `undefined`.
+  webRequestHandler = async ({ requestId, url, redirectUrl }) => {
+    if (this.lastRequestIdReported === requestId) {
+      console.debug(`request ID '${requestId}' already reported, skipping...`);
+      return;
+    }
```

Comment:
You might get interleaved calls so this isn't the best way to protect against double reporting.  I'm thinking that we do not need the listener for onBeforeRequest.  If something redirects in that call, you should still get a call to onBeforeRedirect.